### PR TITLE
feat: multiple tokens per secret

### PR DESCRIPTION
**Pull request by allforabit**
Updates the firebase data model to allow for multiple tokens per secret by creating a `allTokens` top level collection that is has a `secretId` and `subscriptionId` that is optimised for finding a token quickly. The tokens `PUT` endpoint is updated to take an optional existing token and this allows revoking a single token (by deleting the existing token and creating a new one).

A new route has been added to allow deleting secrets. This will delete all tokens that contain the token id. Each delete checks to ensure that the correct subscription id associated with the user is also associated with the token to ensure that they only delete tokens that they are managing.

This PR also fixes the token that is pushed on an OTP request to the correct token. Currently it's sending the expo subscription token.

In addition some improvements were made to the test by using sinon's fake timers to fake the setTimeout and allowed removing the timeout from the test npm script.

### CODE REVIEW

**Comment by simoneb** [comment]
I'm not familiar with advanced data modeling in firestore, but are we sure that duplicating data is the right way to go? In any case, let's make sure we're aware of the implications of using subcollections, i.e. the red warning at the bottom of this page of the docs https://firebase.google.com/docs/firestore/data-model

**Comment by allforabit** [comment]
I've updated now to a much simpler model with just one main collection `allTokens`. I'm adding another route to allow deleting secrets (which will delete all tokens related to a particular secret). The tokens delete route now deletes a particular token and can be used to "revoke" a token. I'm going to implement that now and then fix the tests as well as adding some for  the new route.

**Review by sameer-coder — APPROVED** [review]
Looks good to me

**Comment by allforabit** [comment]
Thanks @sameer-coder I might hold off on merging until this PR lands https://github.com/nearform/optic-expo/pull/222 While the new version is basically working with the existing mobile app, it doesn't tidy up tokens in firebase when the user deletes a token, revokes a token or deletes a secret. The update does happen locally so it's not visible to the user.

File: server/lib/routes/secret.js
Comment (simoneb) [inline server/lib/routes/secret.js:25]: I'm not sure this is secure. Doesn't this mean that as long as we find a record in the subscriptions collection with the provided subscription id and user id, a user can delete ANY token?

File: server/lib/routes/secret.js
Comment (simoneb) [inline server/lib/routes/secret.js:61]: doesn't firestore support a single query to delete multiple records?

File: server/lib/routes/secret.js
Comment (simoneb) [inline server/lib/routes/secret.js:64]: I don't think that a deletion error implies that the user is not authorized

**Review by simoneb — CHANGES_REQUESTED** [review]
I didn't review the whole PR but I have some concerns about the correctness (especially around security) of the current implementation.

I recommend reviewing the code with that in mind.

File: server/lib/routes/secret.js
Comment (allforabit) [reply server/lib/routes/secret.js:25]: The subscription id associated with that token particular token is passed in so if the subscription id of that particular token is not associated with the user it won't progress to deletion.

File: server/lib/routes/secret.js
Comment (allforabit) [reply server/lib/routes/secret.js:61]: The reason I went this route is because each token stores the subscription id so I thought the simplest way was to just iterate over all of these tokens and do the check in `deleteTokenIfAuthorized` to make sure the user can delete that particular item. I do realise that the subscription id is going to be the same for each token associated with a secret so the data is a little duplicated leading to solution I came to here.

**Review by simoneb — APPROVED** [review]

**Pull request merged**
